### PR TITLE
refactor!(chain): Unify ChangeSet nomenclature

### PR DESCRIPTION
**Pull request by danielabrozzoni**
Solves #1022

### Changelog notice

Rename `indexed_tx_graph::IndexedAdditions` to `indexed_tx_graph::ChangeSet`
Rename `indexed_tx_graph::IndexedAdditions::graph_additions` to `indexed_tx_graph::ChangeSet::graph`
Rename `indexed_tx_graph::IndexedAdditions::index_additions` to `indexed_tx_graph::ChangeSet::indexer`
Rename `tx_graph::Additions` to `tx_graph::ChangeSet`
Rename `keychain::DerivationAdditions` to `keychain::ChangeSet`
Rename `CanonicalTx::node` to `CanonicalTx::tx_node`
Rename `CanonicalTx::observed_as` to `CanonicalTx::chain_position`
Rename `LocalChangeSet` to `WalletChangeSet`
Rename `LocalChangeSet::chain_changeset` to `WalletChangeSet::chain`
Rename `LocalChangeSet::indexed_additions` to `WalletChangeSet::indexed_tx_graph`
Rename `LocalUpdate` to `WalletUpdate`
Make `TxGraph::determine_changeset` `pub(crate)`
Add `TxGraph::initial_changeset`
Add `IndexedTxGraph::initial_changeset`
Remove `TxGraph::insert_txout_preview`
Remove `TxGraph::insert_tx_preview`
Remove `insert_anchor_preview`
Remove `insert_seen_at_preview`

### Notes to reviewers

I'm not sure whether we want to keep `TxGraph::determine_changeset` public or not

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

### CODE REVIEW

File: crates/chain/src/tx_graph.rs
Comment (notmandatory) [inline crates/chain/src/tx_graph.rs]: This line makes a less sense with the name change. How about something like:
```suggestion
/// Because [`TxGraph`] is monotone "changeset" can only contain transactions to be added and
```

File: crates/chain/tests/test_tx_graph.rs
Comment (notmandatory) [inline crates/chain/tests/test_tx_graph.rs]: Doesn't this still apply the changeset?
```suggestion
    // Apply changeset and check the new graph counts.
```

**Review by notmandatory — COMMENTED** [review]
LGTM! much easier to read using the same nomenclature across the code base. I added a couple comments/questions. I think this should go in before #1048 and then I'll rebase that PR on these changes.

**Comment by evanlinjin** [comment]
LGTM too. I can ACK and merge after wording changes as mentioned by @notmandatory.

**Comment by LLFourn** [comment]
Haven't reviewed but got a note here: https://github.com/bitcoindevkit/bdk/issues/1060#issuecomment-1669237763 that `TxGraph` doesn't have an "initial_changeset" method to dump the full state as a changeset. It should in order to be consistent. Same with `IndexedTxGraph`. Can we also make sure all changeset types are `must_use` (except for the wallet one).

File: crates/chain/src/indexed_tx_graph.rs
Comment (LLFourn) [inline crates/chain/src/indexed_tx_graph.rs]: It's inside a thing that's a changeset so I'd rather remove this suffix.

```suggestion
    pub graph: tx_graph::ChangeSet<A>,
    /// [`Indexer`] changeset.
    pub indexer: IA,
```

File: crates/chain/src/keychain.rs
Comment (LLFourn) [inline crates/chain/src/keychain.rs]: I suggest renaming the following:

- `chain_changeset` field here to just `chain`
- `indexed_changeset` field here to `indexed_tx_graph` 
- Rename `LocalChangeSet` to `WalletChangeSet`
- Rename `LocalUpdate` to `WalletUpdate`. 

Those last two are suggestions would set the stage a bit for https://github.com/bitcoindevkit/bdk/issues/1061 (I want to move both those types to `bdk/wallet/mod.rs`.

**Review by LLFourn — APPROVED** [review]
Approved as it is but I've made a few extra renaming suggestions and I do think now is the time to create `initial_changeset `methods for the tx graphs. The tx graph one could probably be implemented something like:

```rust
fn initial_changeset(&self) -> ChangeSet {
    Self::default().apply_update(self.clone())
}
```

but for `IndexedTxGraph` you might have to add a `initial_changeset` method to the `Indexer` trait to get it done properly.

**Comment by danielabrozzoni** [comment]
I rebased and fixed the review comments. I'm now working on adding `initial_changeset` for the `tx_graph` and the `indexed_tx_graph`, but since this PR is holding up #1040, feel free to merge before I'm done and I'll open a new PR

**Comment by danielabrozzoni** [comment]
Also, every ChangeSet is `#[must_use]` apart from the local_chain one, but since it's just a type alias, when I tried to add it the compiler warned me that it would have no effect:

```
   Compiling bdk_chain v0.5.0 (/home/daniela/Developer/bitcoindevkit/bdk/crates/chain)
warning: `#[must_use]` has no effect when applied to a type alias
  --> crates/chain/src/local_chain.rs:14:1
   |
14 | #[must_use]
   | ^^^^^^^^^^^
   |
   = note: `#[warn(unused_attributes)]` on by default
```

**Comment by danielabrozzoni** [comment]
Ok, this took less than what I expected :) I just pushed 4ac9c03fc5552026460ded3c1d0b9cea87e01243, adding `initial_changeset` to `tx_graph` and `indexed_tx_graph`.

**Review by notmandatory — APPROVED** [review]
Looks good. ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

File: crates/chain/src/tx_graph.rs
Comment (vladimirfomene) [inline crates/chain/src/tx_graph.rs]: I see a recurrent pattern in these TxGraph methods of calling `determine_changeset` and then `apply_changeset`. Can't we just replace these with `apply_update`?

File: crates/chain/src/indexed_tx_graph.rs
Comment (vladimirfomene) [inline crates/chain/src/indexed_tx_graph.rs]: Do we not need to write self as `self`?

File: crates/chain/src/indexed_tx_graph.rs
Comment (vladimirfomene) [inline crates/chain/src/indexed_tx_graph.rs]: Same as above

File: crates/chain/src/tx_graph.rs
Comment (vladimirfomene) [inline crates/chain/src/tx_graph.rs]: should we not also put self here in ``?

**Review by vladimirfomene — COMMENTED** [review]
Review ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243. Feel free to push the TxGraph refactoring to another PR.

File: crates/chain/src/indexed_tx_graph.rs
Comment (evanlinjin) [inline crates/chain/src/indexed_tx_graph.rs]: Nit: Since we changed the fields of `indexed_tx_graph::ChangeSet`, variable names are now confusing. I think using `changeset.graph` and `changeset.indexer` directly will be better.

File: crates/chain/src/indexed_tx_graph.rs
Comment (evanlinjin) [reply crates/chain/src/indexed_tx_graph.rs]: Nit: This is more descriptive of how this method should be used.

```suggestion
    /// Determines the [`ChangeSet`] between `self` and an empty [`IndexedTxGraph`].
    ///
    /// The returned `ChangeSet` can be applied to an empty `IndexedTxGraph` to recover the current
    /// state of `self`.
```

File: crates/chain/src/tx_graph.rs
Comment (evanlinjin) [inline crates/chain/src/tx_graph.rs]: Nit: I would also make `determine_changeset` consume `update`.

```suggestion
    pub(crate) fn determine_changeset(&self, update: TxGraph<A>) -> ChangeSet<A> {
```

File: crates/chain/src/tx_graph.rs
Comment (evanlinjin) [inline crates/chain/src/tx_graph.rs]: Nit: This might be more efficient as we don't need to apply real changes to `Self::default()`.

```suggestion
        Self::default().determine_changeset(self)
```

**Review by evanlinjin — APPROVED** [review]
ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

Looks good. My comments are regarding documentation, variable naming and minute implementation details (things that are not mission-critical and can be addressed later).

**Comment by notmandatory** [comment]
@danielabrozzoni looks like some small but useful suggestions so I'll wait until you've had a chance to address them before merging.

**Comment by LLFourn** [comment]
ACK 4ac9c03fc5552026460ded3c1d0b9cea87e01243

File: crates/chain/src/tx_graph.rs
Comment (danielabrozzoni) [reply crates/chain/src/tx_graph.rs]: I think we could, but it's better if we do it in another PR, this one is pretty big already :)

**Comment by danielabrozzoni** [comment]
Thanks for the reviews! Updated, and pushed a commit on top to fix the CI

**Review by evanlinjin — APPROVED** [review]
ACK 62f253103c3fd158ad66a0bba11399e7083226f2

**Pull request merged**
